### PR TITLE
Sections new api

### PR DESCRIPTION
**Pull request by Pchelolo**
Takes care of https://phabricator.wikimedia.org/T94890

This version of the PR implements the API from this comment: https://phabricator.wikimedia.org/T143356#2580945

Might need a bit more docs on the format, I'm not quite sure

cc @wikimedia/services

### CODE REVIEW

File: sys/parsoid.js [hunk 1]
```diff
@@ -528,14 +530,46 @@ function replaceSections(original, sectionsJson) {
...
     sectionIds.sort(function(id1, id2) {
         return sectionOffsets[id2].html[0] - sectionOffsets[id1].html[0];
     })
     .forEach(function(id) {
-        var offset = sectionOffsets[id];
-        newBody = newBody.substring(0, offset.html[0])
-        + sectionsJson[id]
-        + newBody.substring(offset.html[1], newBody.length);
+        var sectionReplacement = sectionsJson[id].html;
+        if (typeof sectionReplacement === 'string') {
```
Comment (gwicke) [inline sys/parsoid.js]: I think we can drop strings completely, and always require an array (or undefined) for `html`.

File: sys/parsoid.js [hunk 2]
```diff
@@ -528,38 +530,45 @@ function replaceSections(original, sectionsJson) {
             }
         });
     }
+
+    function sliceSection(id) {
+        var htmlOffset = sectionOffsets[id].html;
+        return originalBody.substring(htmlOffset[0], htmlOffset[1]);
+    }
+
+    function fillSection(id, replacement) {
```
Comment (gwicke) [inline sys/parsoid.js]: Minor: Perhaps `replaceSection` would be a slightly clearer name.

File: sys/parsoid.js [hunk 3]
```diff
@@ -528,38 +530,45 @@ function replaceSections(original, sectionsJson) {
             }
         });
     }
+
+    function sliceSection(id) {
```
Comment (gwicke) [inline sys/parsoid.js]: Altenative name for consideration: `getSectionHTML`

**Comment by Pchelolo** [comment]
@gwicke Updated with new API as discussed

File: test/features/errors/invalid_request.js [hunk 4]
```diff
@@ -103,46 +103,4 @@ describe('400 handling', function() {
...
-            }
-        })
-        .then(function() {
-            throw new Error("400 should be returned");
-        })
-        .catch(function(e) {
-            assert.deepEqual(e.status, 400);
-            assert.contentType(e, 'application/problem+json');
-        });
-    });
```
Comment (Pchelolo) [inline test/features/errors/invalid_request.js]: I've added other tests that check the same

**Comment by gwicke** [comment]
:+1: 

Since this is introducing the API change, we should coordinate merging / deploying with an announcement to the API list.

**Comment by gwicke** [comment]
Announcement sent, with deployment scheduled for Tuesday: 

https://lists.wikimedia.org/pipermail/mediawiki-api-announce/2016-August/000116.html

**Pull request merged**